content/en/docs/Tasks/uninstall-operator.md

```diff
@@ -3,32 +3,45 @@ title: "Uninstall your operator"
 date: 2020-03-25
 weight: 7
 description: >
-  In order to uninstall an operator, you need to delete the following resources:
+  When uninstalling an operator managed by OLM, a Cluster Admin must decide whether or not to remove the CustomResourceDefinitions (CRD), APIServices, and resources related to these types owned by the operator. By design, when OLM uninstalls an operator it does not remove any of the operator's owned CRDs, APIServices, or CRs in order to prevent data loss. Instead, it is left to the Cluster Admin to remove any unwanted types and resources from the cluster. This document will discuss the steps a Cluster Admin should take when uninstalling an operator.
+---
 
-    - Subscription
-    - ClusterServiceVersion (CSV)
+## Step 1: Identifying Resources to Remove
 
-    Both `Subscription` and `CSV` are namespaced objects meaning you need to delete a `Subscription` and a `CSV` in a specific namespace where you install the operator into.
----
+The cluster admin should first understand which types (CRDs and APIServices) are owned by the operator, which is available in the operator's ClusterServiceVersion (CSV) under the `spec.customresourcedefinitions.owned` and `spec.apiservicedefinitions.owned` arrays. It is likely that users have created resources for these types since the operator was installed. The cluster admin should should decide which of these resources to delete on a case-by-case basis. If the resource is not required, delete it. The Cluster Admin should delete all unwanted resources before moving to the next step.
```

Comment:
It might be better to show how admins can inspect the CSV for an operator: 

`kubectl get packagemanifest <operator-name> -o jsonpath="{.spec.customresourcedefinitions.owned}"

Reply:
This command won't work because the path is wrong. The CRD only exists in the `status.channels` array. Example:

```
$ oc get packagemanifests portworx -o yaml
apiVersion: packages.operators.coreos.com/v1
kind: PackageManifest
metadata:
  creationTimestamp: "2020-07-24T13:20:53Z"
  labels:
    catalog: operatorhubio-catalog
    catalog-namespace: olm
    operatorframework.io/arch.amd64: supported
    operatorframework.io/os.linux: supported
    provider: Portworx
    provider-url: ""
  name: portworx
  namespace: operators
  selfLink: /apis/packages.operators.coreos.com/v1/namespaces/operators/packagemanifests/portworx
spec: {}
status:
  catalogSource: operatorhubio-catalog
  catalogSourceDisplayName: Community Operators
  catalogSourceNamespace: olm
  catalogSourcePublisher: OperatorHub.io
  channels:
  - currentCSV: portworx-operator.v1.3.0
    currentCSVDesc:
      annotations:
        alm-examples: |-
          [
            {
              "apiVersion": "core.libopenstorage.org/v1alpha1",
              "kind": "StorageCluster",
              "metadata": {
                "name": "portworx",
                "namespace": "test-operator"
              },
              "spec": {
                "userInterface": {
                  "enabled": true
                },
                "autopilot": {
                  "enabled": false
                }
              }
            },
            {
              "apiVersion": "core.libopenstorage.org/v1alpha1",
              "kind": "StorageNode",
              "metadata": {
                "name": "example",
                "namespace": "test-operator"
              }
            }
          ]
        capabilities: Auto Pilot
        categories: Storage
        certified: "false"
        containerImage: portworx/px-operator:1.3.0
        createdAt: "2020-04-01T08:00:00Z"
        description: Cloud native storage solution for production workloads
        repository: https://github.com/libopenstorage/operator
        support: Portworx, Inc
      apiservicedefinitions: {}
      customresourcedefinitions:
        owned:
        - description: Storage Cluster installs Portworx in the cluster. It has all
            the necessary configurations to setup and update a Portworx cluster.
          displayName: Storage Cluster
          kind: StorageCluster
          name: storageclusters.core.libopenstorage.org
          version: v1alpha1
        - description: DO NOT CREATE Storage Node as it is internally created by the
            operator. It represents the status of a Portworx node.
          displayName: Storage Node
          kind: StorageNode
          name: storagenodes.core.libopenstorage.org
          version: v1alpha1
      description: |
        Portworx-Enterprise is the most widely-used and reliable cloud-native
        storage solution for production workloads and provides high-availability,
        data protection and security for containerized applications.

        Portworx Enterprise enables you to migrate entire applications, including
        data, between clusters in a single data center or cloud, or between clouds,
        with a single kubectl command.

        The cloud native storage and data management platform that enterprises trust
        to manage data in containers now has an operator which simplifies the install,
        configuration, upgrades and manages the Portworx Enterprise cluster lifecycle.

        Learn more about the Portworx Enterprise
        [the data platform for Kubernetes](https://portworx.com/products/introduction)

        To learn more about the platform features, please visit our
        [product features page](https://portworx.com/products/features)

        ### About Portworx

        Portworx is the solution for running stateful containers in production,
        designed with DevOps in mind. With Portworx, users can manage any database
        or stateful service on any infrastructure using any container scheduler,
        including Kubernetes, Mesosphere DC/OS, and Docker Swarm. Portworx solves
        the five most common problems DevOps teams encounter when running stateful
        services in production: persistence, high availability, data automation,
        security, and support for multiple data stores and infrastructure.

        ### How to install StorageCluster

        To customize your cluster's configuration (specification), use the
        [Spec Generator](https://central.portworx.com/) from PX-Central.

        ### Prerequisite

        Ensure ports 17001-17020 on worker nodes are reachable from master and other worker nodes.

        ### Tutorials

        * [Stateful applications on Kubernetes](https://docs.portworx.com/portworx-install-with-kubernetes/application-install-with-kubernetes)

        * [Portworx Enterprise on Kubernetes](https://docs.portworx.com/portworx-install-with-kubernetes)

        * [Kafka on Kubernetes](https://portworx.com/kafka-kubernetes)

        * [Elastisearch on Kubernetes](https://portworx.com/elasticsearch-kubernetes)

        * [PostgreSQL on Kubernetes](https://portworx.com/postgres-kubernetes/)

        * [MongoDB on Kubernetes](https://portworx.com/mongodb-kubernetes/)

        * [Cassandra on Kubernetes](https://portworx.com/cassandra-kubernetes/)

        * [Kubernetes backup and recovery](https://portworx.com/kubernetes-backup/)

        * [Disaster Recovery for Kubernetes](https://portworx.com/kubernetes-disaster-recovery/)

        ### Uninstall

        Deleting the StorageCluster object for Portworx cluster does not stop Portworx
        service running on the nodes, to avoid application downtime.

        To uninstall Portworx completely without wiping the data, you should add the
        following delete strategy to the StorageCluster spec:
        ```
        spec:
          deleteStrategy:
            type: Uninstall
        ```
        **Caution:** To uninstall Portworx and **wipe all the data**, you should use the following
        delete strategy:
        ```
        spec:
          deleteStrategy:
            type: UninstallAndWipe
        ```
      displayName: Portworx Enterprise
      installModes:
      - supported: true
        type: OwnNamespace
      - supported: true
        type: SingleNamespace
      - supported: true
        type: MultiNamespace
      - supported: true
        type: AllNamespaces
      keywords:
      - portworx
      - persistent storage
      - storage
      - cloud native
      - open source
      links:
      - name: Product Features
        url: https://portworx.com/products/features
      - name: Documentation
        url: https://docs.portworx.com
      - name: Support
        url: https://docs.portworx.com/support
      - name: Source Code
        url: https://github.com/libopenstorage/operator
      maintainers:
      - email: support@portworx.com
        name: Portworx
      maturity: stable
      minKubeVersion: 1.12.0
      provider:
        name: Portworx
      relatedImages:
      - portworx/px-operator:1.3.0
      version: 1.3.0
    name: alpha
  - currentCSV: portworx-operator.v1.3.0
    currentCSVDesc:
      annotations:
        alm-examples: |-
          [
            {
              "apiVersion": "core.libopenstorage.org/v1alpha1",
              "kind": "StorageCluster",
              "metadata": {
                "name": "portworx",
                "namespace": "test-operator"
              },
              "spec": {
                "userInterface": {
                  "enabled": true
                },
                "autopilot": {
                  "enabled": false
                }
              }
            },
            {
              "apiVersion": "core.libopenstorage.org/v1alpha1",
              "kind": "StorageNode",
              "metadata": {
                "name": "example",
                "namespace": "test-operator"
              }
            }
          ]
        capabilities: Auto Pilot
        categories: Storage
        certified: "false"
        containerImage: portworx/px-operator:1.3.0
        createdAt: "2020-04-01T08:00:00Z"
        description: Cloud native storage solution for production workloads
        repository: https://github.com/libopenstorage/operator
        support: Portworx, Inc
      apiservicedefinitions: {}
      customresourcedefinitions:
        owned:
        - description: Storage Cluster installs Portworx in the cluster. It has all
            the necessary configurations to setup and update a Portworx cluster.
          displayName: Storage Cluster
          kind: StorageCluster
          name: storageclusters.core.libopenstorage.org
          version: v1alpha1
        - description: DO NOT CREATE Storage Node as it is internally created by the
            operator. It represents the status of a Portworx node.
          displayName: Storage Node
          kind: StorageNode
          name: storagenodes.core.libopenstorage.org
          version: v1alpha1
      description: |
        Portworx-Enterprise is the most widely-used and reliable cloud-native
        storage solution for production workloads and provides high-availability,
        data protection and security for containerized applications.

        Portworx Enterprise enables you to migrate entire applications, including
        data, between clusters in a single data center or cloud, or between clouds,
        with a single kubectl command.

        The cloud native storage and data management platform that enterprises trust
        to manage data in containers now has an operator which simplifies the install,
        configuration, upgrades and manages the Portworx Enterprise cluster lifecycle.

        Learn more about the Portworx Enterprise
        [the data platform for Kubernetes](https://portworx.com/products/introduction)

        To learn more about the platform features, please visit our
        [product features page](https://portworx.com/products/features)

        ### About Portworx

        Portworx is the solution for running stateful containers in production,
        designed with DevOps in mind. With Portworx, users can manage any database
        or stateful service on any infrastructure using any container scheduler,
        including Kubernetes, Mesosphere DC/OS, and Docker Swarm. Portworx solves
        the five most common problems DevOps teams encounter when running stateful
        services in production: persistence, high availability, data automation,
        security, and support for multiple data stores and infrastructure.

        ### How to install StorageCluster

        To customize your cluster's configuration (specification), use the
        [Spec Generator](https://central.portworx.com/) from PX-Central.

        ### Prerequisite

        Ensure ports 17001-17020 on worker nodes are reachable from master and other worker nodes.

        ### Tutorials

        * [Stateful applications on Kubernetes](https://docs.portworx.com/portworx-install-with-kubernetes/application-install-with-kubernetes)

        * [Portworx Enterprise on Kubernetes](https://docs.portworx.com/portworx-install-with-kubernetes)

        * [Kafka on Kubernetes](https://portworx.com/kafka-kubernetes)

        * [Elastisearch on Kubernetes](https://portworx.com/elasticsearch-kubernetes)

        * [PostgreSQL on Kubernetes](https://portworx.com/postgres-kubernetes/)

        * [MongoDB on Kubernetes](https://portworx.com/mongodb-kubernetes/)

        * [Cassandra on Kubernetes](https://portworx.com/cassandra-kubernetes/)

        * [Kubernetes backup and recovery](https://portworx.com/kubernetes-backup/)

        * [Disaster Recovery for Kubernetes](https://portworx.com/kubernetes-disaster-recovery/)

        ### Uninstall

        Deleting the StorageCluster object for Portworx cluster does not stop Portworx
        service running on the nodes, to avoid application downtime.

        To uninstall Portworx completely without wiping the data, you should add the
        following delete strategy to the StorageCluster spec:
        ```
        spec:
          deleteStrategy:
            type: Uninstall
        ```
        **Caution:** To uninstall Portworx and **wipe all the data**, you should use the following
        delete strategy:
        ```
        spec:
          deleteStrategy:
            type: UninstallAndWipe
        ```
      displayName: Portworx Enterprise
      installModes:
      - supported: true
        type: OwnNamespace
      - supported: true
        type: SingleNamespace
      - supported: true
        type: MultiNamespace
      - supported: true
        type: AllNamespaces
      keywords:
      - portworx
      - persistent storage
      - storage
      - cloud native
      - open source
      links:
      - name: Product Features
        url: https://portworx.com/products/features
      - name: Documentation
        url: https://docs.portworx.com
      - name: Support
        url: https://docs.portworx.com/support
      - name: Source Code
        url: https://github.com/libopenstorage/operator
      maintainers:
      - email: support@portworx.com
        name: Portworx
      maturity: stable
      minKubeVersion: 1.12.0
      provider:
        name: Portworx
      relatedImages:
      - portworx/px-operator:1.3.0
      version: 1.3.0
    name: stable
  defaultChannel: stable
  packageName: portworx
  provider:
    name: Portworx
```

Reply:
@awgreene sorry I mean `kubectl get packagemanifest <operator-name> -o jsonpath="{.status.channels[0].currentCSVDesc.customresourcedefinitions.owned}"

Reply:
Does the current CSV always have the same CRDs as the CSV installed on cluster?

Reply:
@awgreene the current CSV has all the information about what's currently installed in the cluster right? 

>The cluster admin should first understand which types (CRDs and APIServices) are owned by the operator, which is available in the operator's ClusterServiceVersion (CSV) under the `spec.customresourcedefinitions.owned` and `spec.apiservicedefinitions.owned` arrays

Which CSV is being referred to here?

Reply:
> @awgreene the current CSV has all the information about what's currently installed in the cluster right?

I believe that the current CSV shows which is at the head of a channel. Tested with the following steps:

1. I created a CatSrc and Subscription with manual approval:
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: busybox-catalog
  namespace: operators
spec:
  displayName: Busybox Catalog
  image: quay.io/agreene/new-busybox-index:1.0.0
  sourceType: grpc
---
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: busybox-sub
  namespace: operators
spec:
  channel: alpha
  installPlanApproval: Manual
  name: busybox
  source: busybox-catalog
  sourceNamespace: operators
```

2. I marked the installplan as approved and installed v1 of the busybox operator version 1.0.0. I then updated the index image to 2.0.0 and the following (unapproved) install plan was created:
```
$ oc get installplans
NAME            CSV              APPROVAL   APPROVED
install-2pbmr   busybox.v2.0.0   Manual     false
install-cdkgk   busybox.v1.0.0   Manual     true

$ oc get csvs
NAME                        DISPLAY              VERSION   REPLACES   PHASE
busybox-dependency.v1.0.0   busybox-dependency   1.0.0                Succeeded
busybox.v1.0.0              busybox              1.0.0                Succeeded
```

This shows that 2.0.0 is available on the cluster, but 1.0.0 is installed. The output from the packagemanifest shows 2.0.0, not the installed version:

```
$ oc get packagemanifests.packages.operators.coreos.com busybox -o yaml
apiVersion: packages.operators.coreos.com/v1
kind: PackageManifest
metadata:
  creationTimestamp: "2020-07-28T15:10:37Z"
  labels:
    catalog: busybox-catalog
    catalog-namespace: operators
    operatorframework.io/arch.amd64: supported
    operatorframework.io/os.linux: supported
    provider: Red Hat
    provider-url: ""
  name: busybox
  namespace: operators
  selfLink: /apis/packages.operators.coreos.com/v1/namespaces/operators/packagemanifests/busybox
spec: {}
status:
  catalogSource: busybox-catalog
  catalogSourceDisplayName: Busybox Catalog
  catalogSourceNamespace: operators
  catalogSourcePublisher: ""
  channels:
  - currentCSV: busybox.v2.0.0
    currentCSVDesc:
      annotations:
        olm.skipRange: '>=0.0.0 <2.0.0'
      apiservicedefinitions: {}
      customresourcedefinitions:
        required:
        - description: Foo resources for testing dependencies
          displayName: Foo
          kind: Foo
          name: foos.olm.test.io
          version: v1
      description: |
        A busybox CSV.
      displayName: busybox
      installModes:
      - supported: true
        type: OwnNamespace
      - supported: true
        type: SingleNamespace
      - supported: true
        type: MultiNamespace
      - supported: true
        type: AllNamespaces
      maturity: alpha
      provider:
        name: Red Hat
      relatedImages:
      - busybox
      version: 2.0.0
    name: alpha
  defaultChannel: alpha
  packageName: busybox
  provider:
    name: Red Hat
```

Reply:
> Which CSV is being referred to here?

@anik120 this would be the installed CSV on the cluster.